liberty-managed/src/main/java/io/openliberty/arquillian/managed/WLPManagedContainer.java

```diff
@@ -909,112 +924,82 @@ private void logAllApps() {
+          
+          for (AppStateChecker appStateChecker : appStateCheckers) {
+              String currentState = appStateChecker.checkState();
+              
+              if (log.isLoggable(FINER)) {
+                  log.finer("AppMBean for " + appStateChecker.getAppName() + " is in state " + currentState);
+              }
+              
+              if (!currentState.equals(desiredState)) {
+                  allReady = false;
+              }
+              
+              // Applications should only be in INSTALLED state for a very short time before moving to STARTING
+              // An application which stays in INSTALLED state has failed to start
+              if (desiredState.equals("STARTED") && currentState.equals("INSTALLED") && appStateChecker.getMsInState() >= 1000) {
+                  log.finer(appStateChecker.getAppName() + " has failed to start (in INSTALLED state for more than one second)");
+                  throwDeploymentException(appStateChecker.getAppName());
+              }
+          }
+          
+          if (allReady) {
+              break;
+          }
+          try {
+              Thread.sleep(100);
+          } catch (InterruptedException e) {
+              throw new DeploymentException("Deployment interrupted", e);
+          }
+          timeleft -= 100;
       } while (timeleft > 0);
-
-      // If we haven't returned in the while loop, not all apps were ready in the given timeout period.
-
-      logAllApps();
-
-      String appMessageStatus = "";
-      for(Entry<ObjectName, AppStatus> entry : appMBeans.entrySet()) {
-         // Timeout while waiting for ApplicationMBean to reach targetState
-         String appName = entry.getKey().getCanonicalName();
-         AppStatus status = entry.getValue();
-         if(status == AppStatus.INITIAL) {
-            appMessageStatus += "Timeout while waiting for \"" + appName + "\" ApplicationMBean to reach targetState.\n";
-         }
-         else if(status == AppStatus.MATCHES_TARGET_STATE) {
-            appMessageStatus += "Timeout while waiting for \"" + appName + "\" ApplicationState to reach STARTED.\n";
-         }
+      
+      if (!allReady) {
+          // We're about to get the status of apps one more time to build the error message
+          // but there's a possibility they've all finished between now and the last time we checked
+          // so we reset allReady and compute it again as we loop through the apps
+          allReady = true;
+          
+          StringBuilder appMessageStatus = new StringBuilder();
+          for (AppStateChecker appStateChecker : appStateCheckers) {
+              if (!appStateChecker.checkState().equals(desiredState)) {
+                  appMessageStatus.append("Timeout while waiting for \"")
+                          .append(appStateChecker.getAppName())
+                          .append("\" ApplicationMBean to reach ")
+                          .append(desiredState);
```

Comment:
I can't see where it logs the actual state (what used to be "status") at the error level or in then exception. It is logged as currentState in line 951 but at a logging level that is less likely to be operative when an unexpected problem occurs?? [Unless one thinks of Arquillian being used for test only and thus mostly at high log levels???)

Reply:
I agree, I'll add that